pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -109,35 +84,147 @@ class BudgetOptimizer(BaseModel):
+        self._pymc_model = self.mmm_model._set_predictors_for_optimization(
+            self.num_periods
+        )
+        self._budget_dims = [
+            dim
+            for dim in self._pymc_model.named_vars_to_dims["channel_data"]
+            if dim != "date"
+        ]
+        self._budget_coords = {
+            dim: list(self._pymc_model._coords[dim]) for dim in self._budget_dims
+        }
+        self._budget_shape = tuple(len(coord) for coord in self._budget_coords.values())
+
         self._compiled_functions = {}
         self._compile_objective_and_grad()
 
-    def _compile_objective_and_grad(self):
-        """Compile the objective function and its gradient using symbolic computation."""
-        budgets_sym = pt.vector("budgets")
+    def _create_budget_variable(self) -> pt.TensorVariable:
+        size_budgets = np.prod(self._budget_shape)
+        budgets_flat = pt.tensor("budgets_flat", shape=(size_budgets,))
+        budgets = budgets_flat.reshape(self._budget_shape)
+        return budgets
+
+    def extract_response_distribution(
+        self, budgets: pt.TensorVariable
+    ) -> pt.TensorVariable:
+        """Extract the response graph, conditioned on the posterior draws and a placeholder budget variable."""
+        if not (isinstance(budgets, pt.TensorVariable)):  # and budgets.type.ndim == 1):
+            raise ValueError("budgets must be a TensorVariable")
+
+        num_periods = self.num_periods
+        model = self._pymc_model
+        posterior = self.mmm_model.idata.posterior  # type: ignore
+        max_lag = self.mmm_model.adstock.l_max
+        channel_scales = self.mmm_model._channel_scales
+
+        # Freeze all but date dims for a more succinct graph
+        coords = self._pymc_model._coords
+        model = freeze_dims_and_data(
+            model, data=[], dims=[dim for dim in coords if dim != "date"]
+        )
+
+        response_variable = model[self.response_variable]
+
+        # Replicate the budget over num_periods and append zeros to also quantify carry-over effects
+        channel_data_dims = model.named_vars_to_dims["channel_data"]
+        date_dim_idx = list(channel_data_dims).index("date")
+
+        budgets_tiled_shape = list(tuple(budgets.shape))
+        budgets_tiled_shape.insert(date_dim_idx, num_periods)
+        # TODO: If scales become part of the model, we don't need to transform it here
+        budgets /= channel_scales
+        budgets_tiled = pt.broadcast_to(
+            pt.expand_dims(budgets, date_dim_idx), budgets_tiled_shape
+        )
+
+        budget_full_shape = list(tuple(budgets.shape))
+        budget_full_shape.insert(date_dim_idx, num_periods + max_lag)
+        budgets_full = pt.zeros(budget_full_shape)
+        set_idxs = (*((slice(None),) * date_dim_idx), slice(None, num_periods))
+        budgets_full = budgets_full[set_idxs].set(budgets_tiled)
+        budgets_full.name = "budgets_full"
```

Comment:
these lines are a bit hard to read / digest . Shall we extract the to simple functions (small suggestion)

Reply:
Done